File: UVa/Problem Solving/DP/UVa_10003_Cutting_Sticks.py
import sys
sys.stdin = open('input.txt')


def solution(n, cuts):
    memo = [[-1] * 60 for _ in range(60)]

    def search(left, right):
        if left + 1 == right:
            return 0
        if memo[left][right] == -1:
            memo[left][right] = 2e9
            for idx in range(left + 1, right):
                memo[left][right] = min(memo[left][right], search(
                    left, idx) + search(idx, right) + cuts[right] - cuts[left])
        return memo[left][right]

    return search(0, n + 1)


inp = iter(sys.stdin.readlines())
while True:
    l = int(next(inp))
    if not l:
        break
    n = int(next(inp))
    cuts = list(map(int, next(inp).split()))
    cuts = [0] + cuts + [l]
    print(f"The minimum cutting is {solution(n, cuts)}.")

File: UVa/Problem Solving/DP/UVa_10003_Cutting Sticks.cpp
// https://onlinejudge.org/external/100/10003.pdf

#include <bits/stdc++.h>
using namespace std;
#define FOR(i, a, b) for (int i = a; i < b; ++i)
#define RESET(a, b) memset(a, b, sizeof(a))
#define INF 1000000000

int length, n;
int c[50];

int memo[1000][1000];

int cut(int left, int right)
{
    if (left + 1 == right)
        return 0;
    int &ans = memo[left][right];
    if (ans != -1)
        return ans;
    ans = INF;
    FOR(i, left + 1, right)
    {
        ans = min(ans, c[right] - c[left] + cut(left, i) + cut(i, right));
    }
    return ans;
}

int main()
{
    c[0] = 0;
    while (scanf("%d", &length), length)
    {
        RESET(memo, -1);
        scanf("%d", &n);
        FOR(i, 1, n + 1)
        {
            scanf("%d", &c[i]);
        }
        c[n + 1] = length;

        printf("The minimum cutting is %d.\n", cut(0, n + 1));
    }
    return 0;
}